src/v/storage/disk_log_impl.h

```diff
@@ -274,6 +283,13 @@ class disk_log_impl final : public log {
 
     // Mutually exclude operations that will cause segment rolling
     // do_housekeeping and maybe_roll
+    //
+    // This lock will only rarely be contended.  If it is held, then
+    // we must wait for do_housekeeping to complete before proceeding, because
+    // the log might be in a state mid-roll where it has no appender.
+    // We need to take this irrespective of whether we're actually rolling
+    // or not, in order to ensure that writers wait for a background roll
+    // to complete if one is ongoing.
```

Comment:
nit: this lock is also grabbed when performing truncation. That's great since that will release the appender too and the write path needs protection. Perhaps we could update this comment to reflect this extra fact.

Reply:
Done